src/geometry/manifold/manifold.lean

```diff
@@ -0,0 +1,539 @@
+  comp := λe e' he he', begin
+    cases he; simp at he he',
+    { simpa [he] },
+    { have : (e ≫ₕ e').source ⊆ e.source := sep_subset _ _,
+      rw he at this,
+      have : (e ≫ₕ e') ∈ {e : local_homeomorph H H | e.source = ∅} := disjoint_iff.1 this,
+      exact (mem_union _ _ _).2 (or.inr this) },
+  end,
+  inv := λe he, begin
+    have E := (mem_union _ _ _).1 he,
+    simp at E,
+    cases E,
+    { rw E, finish },
+    { have : e.target = ∅,
+      { rw [← e.to_local_equiv.image_source_eq_target, E],
+        erw image_empty },
+      have : e.symm ∈ {e : local_homeomorph H H | e.source = ∅} := this,
+      exact (mem_union _ _ _).2 (or.inr this) },
+  end,
+  id_mem := mem_union_left _ (mem_insert _ ∅),
+  locality := λe he, begin
+    by_cases h : e.source = ∅,
+    { apply mem_union_right, exact h },
+    { rcases ne_empty_iff_exists_mem.1 h with ⟨x, hx⟩,
+      rcases he x hx with ⟨s, open_s, xs, hs⟩,
+      have x's : x ∈ (e.restr s).source,
+      { rw [restr_source, interior_eq_of_open open_s],
+        exact ⟨hx, xs⟩ },
+      cases hs,
+      { simp at hs,
+        have : (e.restr s).source = univ, by { rw hs, simp },
+        simp at this,
+        have : univ ⊆ interior s, by { rw ← this, exact inter_subset_right _ _ },
+        have : s = univ, by rwa [interior_eq_of_open open_s, univ_subset_iff] at this,
+        rw [this, restr_univ] at hs,
+        rw hs,
+        exact mem_union_left _ (mem_insert _ _) },
+      { rw mem_set_of_eq at hs,
+        rw hs at x's,
+        exfalso,
+        exact x's } },
+  end,
+  eq_on_source := λe e' he he'e, begin
+    cases he,
+    { have : e = e',
+      { refine eq_of_eq_on_source_univ (setoid.symm he'e) _ _;
+        rw set.mem_singleton_iff.1 he; refl },
+      rw ← this,
+      apply mem_union_left _ he },
+    { simp at he,
+      have : e'.source = e.source := source_eq_of_eq_on_source he'e,
+      have : e'.source = ∅, by { rw this, exact he },
+      apply mem_union_right,
+      exact this }
+  end }
+
+/-- Every structure groupoid contains the identity groupoid -/
+instance : lattice.order_bot (structure_groupoid H) :=
+{ bot    := id_groupoid H,
+  bot_le := begin
+    assume u f hf,
+    change f ∈ {local_homeomorph.refl H} ∪ {e : local_homeomorph H H | e.source = ∅} at hf,
+    simp only [singleton_union, mem_set_of_eq, mem_insert_iff] at hf,
+    cases hf,
+    { rw hf,
+      apply u.id_mem },
+    { apply u.locality,
+      assume x hx,
+      rw [hf, mem_empty_eq] at hx,
+      exact hx.elim }
+  end,
+  ..structure_groupoid.partial_order }
+
+/-- To construct a groupoid, one may consider classes of local homeos such that both the function
+and its inverse have some property. If this property is stable under composition,
+one gets a groupoid. `pregroupoid` bundles the properties needed for this construction, with the
+groupoid of smooth functions with smooth inverses as an application. -/
+structure pregroupoid (H : Type*) [topological_space H] :=
```

Comment:
Yes, we would, but since I am essentially mimicking groupoids (the goal is really to construct groupoids) and since groupoids are bundled, it seems coherent to also use the fully bundled version here. (And this gives rise to less heavy notation when one uses it, as one does not need to repeat the property all the time).